local/generic ssh driver for minikube

dims opened the pull request:
Just scp the minikube into VM or host and run:
out/minikube --v=9 start --logtostderr --vm-driver=local

* We do not need the minikube ISO, so don't download it.
* Use 127.0.0.1 as the IP
* Get the user name and the SSH id_rsa from the environment.
* Support only Linux for the moment.
* Initial code picked up from:
  https://github.com/docker/machine/blob/master/drivers/generic/generic.go

Earlier and ongoing discussions, point to overriding SSH in the
libmachine layer somehow. This PR takes the approach of of reusing
the information already available in the environment to support
the no-vm use case.

Note that it would be easy to supply the IP, id_rsa via command line
to even support a remote host/vm case if necessary.

r2d4 commented on pkg/minikube/cluster/cluster_linux.go:
```diff
@@ -21,7 +21,10 @@ import (
 	"path/filepath"
 
 	kvm "github.com/dhiltgen/docker-machine-kvm"
+	local "github.com/dims/docker-machine-local"
```
Whats the difference between your driver and libmachine's?

Either way, we should have the driver code in minikube

dims replied:
@r2d4 : Ack, will drop my repo and copy the code directly into minikube. Since we don't have https://github.com/docker/machine/blob/master/drivers/generic/generic.go in our vendor'ed stuff. The difference is minimal and once we have that vendored, we can get rid of the local.go.

r2d4 commented on pkg/minikube/machine/client_linux.go:
```diff
@@ -41,11 +44,23 @@ func getKVMDriver(rawDriver []byte) (drivers.Driver, error) {
 	switch driverName {
 	case "virtualbox":
 		plugin.RegisterDriver(virtualbox.NewDriver("", ""))
+	case "local":
```
We shouldn't support the RPC-separate binary client for the local driver.  The LocalClient allows the libmachine API to access the driver directly, and this should be the only way that the local client is used.  So no need to register the driver here.

dims replied:
Ack removing it.

r2d4 commented on pkg/minikube/cluster/cluster.go:
```diff
@@ -292,8 +292,10 @@ func createVirtualboxHost(config MachineConfig) drivers.Driver {
-	if err := config.Downloader.CacheMinikubeISOFromURL(config.MinikubeISO); err != nil {
-		return nil, errors.Wrap(err, "Error attempting to cache minikube ISO from URL")
+	if config.VMDriver != "local" {
+		if err := config.Downloader.CacheMinikubeISOFromURL(config.MinikubeISO); err != nil {
```
the downloader is an interface, and I think it might be a little cleaner to supply a no-op implementation earlier in the code, in cmd/start where its passed into machine config.  Although this seems ok to me, I wonder if its easy enough to change it in cmd

dims replied:
ack, let's get this version in and i can then iterate to make it cleaner.

dlorenc commented on pkg/minikube/cluster/cluster_non_linux_panic.go:
```diff
@@ -23,3 +23,7 @@ import "github.com/docker/machine/libmachine/drivers"
+
+func createLocalHost(config MachineConfig) drivers.Driver {
+	panic("local not supported")
+}
```
nit: add newline.

dims replied:
Done

dlorenc commented on pkg/minikube/cluster/local/local.go:
```diff
@@ -0,0 +1,201 @@
+
+// GetCreateFlags registers the flags this driver adds to
+// "docker hosts create"
+func (d *Driver) GetCreateFlags() []mcnflag.Flag {
```
We'll never actually go through these codepaths in minikube, so feel free to leave these functions blank (GetCreateFlags, SetConfigFromFlags).

dims replied:
Done

dlorenc commented:
@minikube-bot OK to test

dims commented:
whoa! all tests passed thanks @dlorenc Also if you can point me to some info on how to add a "Linux-Local" i can take care of that either in this PR or on another PR

r2d4 commented:
@dims you'll need to add a jobname here https://github.com/kubernetes/minikube/blob/master/hack/jenkins/minikube_set_pending.sh#L30

and create a script like https://github.com/kubernetes/minikube/blob/master/hack/jenkins/linux_integration_tests_virtualbox.sh

in addition, it still requires us to manually set up the job in our internal jenkins, which I can do for you

dims commented:
@r2d4 - Thanks! i added/updated as you mentioned. added some debug information/steps to make sure local driver will work in the jenkins environment. PTAL when you get a chance. No rush.

r2d4 commented:
I set up the jenkins job for linux-local, so it should get triggered on a retest

r2d4 commented:
@minkube-bot retest this please

r2d4 commented:
@minikube-bot retest this please

dims commented:
Thanks @r2d4 Matt

dims commented:
@r2d4 Matt, the jenkins jobs are not triggering. Latest run on Linux-Local does not reflect my latest changes.

dims commented:
@r2d4 Oh, just to give you an idea where i am stuck... I need to be able to run sudo (without failing on tty/pass) to get the driver working

dims commented:
disabling the Linux-Local with a NOTE in the script to add it back later when CI supports password-less sudo

omeid commented:
Travis supports password less sudo, you just need to add `sudo: required` at the root of `.travis-ci`.
